### PR TITLE
Tjanc/purge refract

### PR DESCRIPTION
**Pull request by tjanc**
Redefines Data Structure Elements without any references to Refract.

### CODE REVIEW

File: docs/element-definitions.md [hunk 1]
```diff
@@ -1,1419 +1,1612 @@
...
 
-            Type attributes of a type definition.
+Below is an example of how a profile link is used as a meta link.
 
-            Note, if `sample` (or `default`) attribute is specified the value SHOULD be stored in the `samples` (or `default`) property instead of the element's `content`.
+```json
+{
+  "element": "foo",
+  "meta": {
+    "links": [
```
Comment (kylef) [inline docs/element-definitions.md]: This isn't valid JSON serialisation, the links should be an array element, meta values MUST be elements themselves.

```json
{
    "element": "array",
    "content": [
        { "element": "link", ... }
    ]
}
```

File: docs/element-definitions.md [hunk 2]
```diff
@@ -1,1419 +1,1612 @@
...
+### Null Element
+Type with domain of a single value. Called `unit` or `()` in some programming languages.
+Note that both `nullptr` and `void` from C-style languages are unrelated concepts.
 
-The value of `Templated Href` type is to be used as a URI Template, as defined in [RFC 6570][].
-The value of the `Templated Href` type is a template used to determine the target URI of the related resource or transition.
-The value of the `Templated Href` type SHOULD be resolved as a URI-Reference per [RFC 3986][] and MAY be a relative reference to a URI.
+#### Template
+- `element` - `"null"`
+- `content` - MUST hold unit if set
```
[the remark this answers is not in the document]
Comment (tjanc) [reply docs/element-definitions.md]: This just says the content can only be set to `unit`, whatever that means in the target language. For example, in JSON, `unit` might be `null`.

[the remark this answers is not in the document]
Comment (tjanc) [reply docs/element-definitions.md]: Suggestions for re-wording?

[the remark this answers is not in the document]
Comment (kylef) [reply docs/element-definitions.md]: I don't think we serialise null elements with content anywhere. I think content should be omitted.

File: docs/element-definitions.md [hunk 3]
```diff
@@ -1,1419 +1,1623 @@
...
+}</pre>
+      </td>
+      <td>
+        <pre>{}</pre>
+        <pre>{
+  "foo": "Hey!"
+}</pre>
+        <pre>{
+  "bar": true
+}</pre>
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's separate out the member element from this. For the example value of the member element, we should say that there is no JSON equivalent without the context but we can show them the part of an object that it represents. `"key": value`

File: docs/element-definitions.md [hunk 4]
```diff
@@ -1,1419 +1,1623 @@
...
+        <pre>{}</pre>
+        <pre>{
+  "foo": "Hey!"
+}</pre>
+        <pre>{
+  "bar": 42.3
+}</pre>
+        <pre>{
+  "foo": "Hey!",
+  "bar": 42.3,
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's change the values of `foo` and `bar` in the above 2 lines of this comment.

File: docs/element-definitions.md [hunk 5]
```diff
@@ -1,1419 +1,1623 @@
...
+  </thead>
+  <tbody>
+    <tr>
+      <td><a href="#ref-element">ref</a></td>
+      <td>
+        Given an named type such as
+        <pre>{
+  "element": "string",
+  "meta": {
+    "id": "SpecialString"
```
Comment (pksunkara) [inline docs/element-definitions.md]: `id` should be a string element, not a string.

File: docs/element-definitions.md [hunk 6]
```diff
@@ -1,1419 +1,1623 @@
...
 
-#### Properties
+---
 
-- `element`: hrefVariables (string, fixed)
+### Boolean Element
 
-### Data Structure (Base API Element)
+Type with domain of two values: true and false.
+`content` property MUST contain a boolean value if set.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's remove the second sentence since it's redundant.

File: docs/element-definitions.md [hunk 7]
```diff
@@ -1,1419 +1,1623 @@
...
 
-Data structure definition using Data Structure elements.
+#### Template
+- `element` - `"boolean"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - domain restricted to value given in `content`
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Boolean]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (Boolean Element) - default value for this Element; type of `default` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Boolean instead of Boolean Element. And it should be linked.

File: docs/element-definitions.md [hunk 8]
```diff
@@ -1,1419 +1,1623 @@
...
+- `content` - ⊥ (false) or ⊤ (true)
```
Comment (pksunkara) [inline docs/element-definitions.md]: Remove formal notation and let's just say "false or true"

File: docs/element-definitions.md (hunk 7 above)
Comment (pksunkara) [reply docs/element-definitions.md]: type of `default` MUST match the type this element describes. (Everywhere else where you use this sentence)

File: docs/element-definitions.md [hunk 9]
```diff
@@ -1,1419 +1,1623 @@
...
+`content` property MUST contain a boolean value if set.
 
-Data structure definition using Data Structure elements.
+#### Template
+- `element` - `"boolean"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - domain restricted to value given in `content`
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Boolean]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
```
[the remark this answers is not in the document]
Comment (pksunkara) [reply docs/element-definitions.md]: type of items in `samples` MUST match the type this element describes (Everywhere else where you use this sentence)

[the remark this answers is not in the document]
Comment (pksunkara) [reply docs/element-definitions.md]: Let's omit `array of`. ((Everywhere else where you use this sentence))

File: docs/element-definitions.md [hunk 10]
```diff
@@ -1,1419 +1,1623 @@
...
-### Data Structure (Base API Element)
+Type with domain of two values: true and false.
+`content` property MUST contain a boolean value if set.
 
-Data structure definition using Data Structure elements.
+#### Template
+- `element` - `"boolean"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - domain restricted to value given in `content`
```
Comment (pksunkara) [inline docs/element-definitions.md]: The type this element describes is restricted to value given in `content`. (Everywhere else where you use this sentence for primitives)

File: docs/element-definitions.md [hunk 11]
```diff
@@ -1,1419 +1,1623 @@
...
-        The `content` MUST NOT include more than one `Data Structure`.
+
+#### Template
+- `element` - `"number"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - domain restricted to value given in `content`
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Number]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (Number) - default value for this Element; type of `default` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Bad linking for `samples` and `default`

File: docs/element-definitions.md [hunk 12]
```diff
@@ -1,1419 +1,1623 @@
...
+- `content` - rational number
 
 #### Example
 
+Type Element matching only rationals. Matches JSON `number`.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Matches JSON number values.

File: docs/element-definitions.md [hunk 13]
```diff
@@ -1,1419 +1,1623 @@
...
-- `element`: transition (string, fixed)
+#### Template
+- `element` - `"string"`
 - `attributes`
-    - `relation` - ([String Element][]) - Link relation type as specified in [RFC 5988][].
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - Domain restricted to value given in `content`.
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[String]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (String) - default value for this Element; type of `default` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Bad linking for `samples` and `default`

File: docs/element-definitions.md [hunk 14]
```diff
@@ -1,1419 +1,1623 @@
...
+- `content` - finite character string
 
-        The value of `relation` attribute SHOULD be interpreted as a link relation
-        between transition's parent resource and the transition's target resource
-        as specified in the `href` attribute.
+#### Example
 
-    - `href` (Templated Href) - The URI template for this transition.
+Type Element matching only character strings. Matches a `string` in JSON.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Matches JSON string values.

File: docs/element-definitions.md [hunk 15]
```diff
@@ -1,1419 +1,1623 @@
...
 
-        If present, the value of the `href` attribute SHOULD be used to resolve
-        the target URI of the transition.
+Because the definition of an Element does not depend on syntax, we might serialize it into JSON:
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's remove this line.

File: docs/element-definitions.md [hunk 16]
```diff
@@ -1,1419 +1,1623 @@
...
-  "content": []
-}
-```
+---
 
-### API Metadata (Member Element)
+### Array Element
 
-#### Properties
+Type with domain of all finite lists.
```
Comment (pksunkara) [inline docs/element-definitions.md]: finite lists of values

File: docs/element-definitions.md [hunk 17]
```diff
@@ -1,1419 +1,1623 @@
...
 
-### Category (Base API Element)
+#### Template
+- `element` - `"array"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - Restricts domain to a positionally typed fixed-length list over types in content. This type class is usually called a <i>tuple</i>, <i>sum type</i> or <i>Π-type</i>.
+    - `fixedType` ([String](#string-element)) - Restricts domain to a list of types given in `content`.
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Array]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Bad linking

File: docs/element-definitions.md [hunk 18]
```diff
@@ -1,1419 +1,1623 @@
...
-                - `content` (enum)
-                    - user (string) - User-specific metadata. Metadata written in the source.
-                    - adapter (string) - Serialization-specific metadata. Metadata provided by adapter.
 
-### Category (Base API Element)
+#### Template
+- `element` - `"array"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - Restricts domain to a positionally typed fixed-length list over types in content. This type class is usually called a <i>tuple</i>, <i>sum type</i> or <i>Π-type</i>.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's replace the last sentence with something like:

It further fixes the content of the items given that they declare a value.

File: docs/element-definitions.md [hunk 19]
```diff
@@ -1,1419 +1,1623 @@
...
+    },
+    {
+      "element": "number"
+    }
+  ]
+}
+
+```
+
+Type Element matching only lists of one of `string`, `number`.
```
Comment (pksunkara) [inline docs/element-definitions.md]: lists of items where each item is either a `string` or a `number`

File: docs/element-definitions.md [hunk 20]
```diff
@@ -1,1419 +1,1623 @@
...
+An _Element_ SHALL be a tuple (`element`, `meta`, `attributes`, `content`) where
+ - `element` SHALL be a non-empty, finite character string identifying the _type_ of this Element
+ - `meta` SHALL be a set of _properties_, some of which have [reserved semantics](#reserved-meta-properties)
+ - `attributes` SHALL be a set of _properties_ defined by the _type_ of this Element
+ - `content` SHALL be defined by the _type_ of this Element
 
-The API Elements reference relies on [Refract][] for its definition and structure. To make this reference document more understandable, this base element has been included and used throughout.
 
-This base element defines the structure of each element in this reference. Elements then extend upon this structure in their own definitions throughout.
+A _property_ SHALL be a tuple (`key`, `value`) where
```
[the remark this answers is not in the document]
Comment (tjanc) [reply docs/element-definitions.md]: I disagree. In JSON & JS it's called a _property_.

File: docs/element-definitions.md [hunk 21]
```diff
@@ -1,1419 +1,1623 @@
...
+#### Template
+- `element` - `"object"`
 - `attributes`
-    - `contentType` ([String Element][]) - Optional media type of the content.
-- `content` (string)
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - SHALL restrict the domain to a specific instance of a property list. Entries in `content` of this Object Element MUST contain at least one of content, default and sample.
+    - `fixedType` ([String](#string-element)) - SHALL restrict the domain to properties listed in `content`.
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Object]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Bad linking for `samples` and `default`

File: docs/element-definitions.md [hunk 22]
```diff
@@ -1,1419 +1,1623 @@
...
 
-- `element`: copy (string, fixed)
+#### Examples
+See [Object Element](#object-element) for examples.
+
+---
+
+### Object Element
+
+Type with domain of all finite lists of _properties_.
```
Comment (pksunkara) [inline docs/element-definitions.md]: unordered list

File: docs/element-definitions.md [hunk 23]
```diff
@@ -1,1419 +1,1623 @@
...
+> - two properties SHALL be equal if their keys are.
+
+#### Template
+- `element` - `"object"`
 - `attributes`
-    - `contentType` ([String Element][]) - Optional media type of the content.
-- `content` (string)
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - SHALL restrict the domain to a specific instance of a property list. Entries in `content` of this Object Element MUST contain at least one of content, default and sample.
+    - `fixedType` ([String](#string-element)) - SHALL restrict the domain to properties listed in `content`.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Restricts the domain to a fixed-length unordered list of properties and makes properties implicitly required

File: docs/element-definitions.md [hunk 24]
```diff
@@ -1,1419 +1,1623 @@
...
+> - `value` SHALL be an _Element_
+> - two properties SHALL be equal if their keys are.
+
+#### Template
+- `element` - `"object"`
 - `attributes`
-    - `contentType` ([String Element][]) - Optional media type of the content.
-- `content` (string)
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - SHALL restrict the domain to a specific instance of a property list. Entries in `content` of this Object Element MUST contain at least one of content, default and sample.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Restricts the domain to a fixed-length unordered list of properties, making them implicitly required. It further applies fixed to an item given it declares content or default.

File: docs/element-definitions.md [hunk 25]
```diff
@@ -1,1419 +1,1623 @@
...
+  - `validation` - _reserved for future use_
+  - `samples` ([Array[Object]](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (Object) - default value for this Element; type of `default` MUST match the type of this Element
+- `content` - list of any of
+  - [Member](#member-element) - object property
+  - [Extend](#extend-element) - MUST type a property
+  - [Select](#select-element) - contained [Option Element](#option-element)s MUST type properties
+  - [Ref](#ref-element) - MUST reference an Object Element
+
+[References](#ref-element) in the `content` of an Object Element [Ref Element](#ref-element) SHALL be semantically equivalent to their substitution by items held in the content of the referenced Object Element. Less formally, Ref Elements in the content of Object Element represent in-place mixins.
```
Comment (pksunkara) [inline docs/element-definitions.md]: The occurence of `Ref Element` should be removed along with the link.

File: docs/element-definitions.md [hunk 26]
```diff
@@ -1,1419 +1,1623 @@
...
-      "element": "copy",
-      "content": "Lorem Ipsum"
+      "element": "member",
+      "content": {
+        "key": {
+          "element": "string",
+          "content": "foo"
+        },
+        "value": {
+          "element": "boolean"
```
Comment (pksunkara) [inline docs/element-definitions.md]: comma at the end

File: docs/element-definitions.md [hunk 27]
```diff
@@ -1,1419 +1,1623 @@
...
-      "element": "resource"
-    },
+      "element": "member",
+      "content": {
+        "key": {
+          "element": "string",
+          "content": "foo"
+        },
+        "value": {
+          "element": "string"
```
Comment (pksunkara) [inline docs/element-definitions.md]: comma at the end

File: docs/element-definitions.md [hunk 28]
```diff
@@ -1,1419 +1,1623 @@
...
+          "content": "fun"
+        }
+      }
+    }
+  ]
+}
+
+```
+
+Type Element matching only a property list with key "foo" of value type `boolean` and with the key "bar" of value type `string` (JSON `{"foo": false, "bar": "fun"}`, `{"foo": true, "bar": ""}`).
```
Comment (pksunkara) [inline docs/element-definitions.md]: type `etc..` at the end of the examples.

File: docs/element-definitions.md [hunk 29]
```diff
@@ -1,1419 +1,1623 @@
...
+> - `key` SHALL be a non-empty, finite character string
+> - `value` SHALL be an _Element_
+> - two properties SHALL be equal if their keys are.
 
-#### Properties
+#### Template
+- `element` - `"member"`
+- `attributes`
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `required` ([String](#string-element)) - Property MUST be present in Object containing it. I.e. restricts the domain of the containing Object Element type to one containing this property.
```
Comment (pksunkara) [inline docs/element-definitions.md]: `present in value represented by the containing [Object Element](#object-element)`

File: docs/element-definitions.md [hunk 30]
```diff
@@ -1,1419 +1,1623 @@
...
+    - `optional` ([String](#string-element)) - Property MAY NOT be present in Object containing it. I.e. expands the domain of the containing Object Element type to one not containing this property.</p>
```
Comment (pksunkara) [inline docs/element-definitions.md]: `present in value represented by the containing [Object Element](#object-element)`

File: docs/element-definitions.md [hunk 31]
```diff
@@ -1,1419 +1,1623 @@
...
 
-#### HTTP Transaction (Base API Element)
+---
 
-Example of an HTTP Transaction. The example's content consist of a request-response
-message pair. A transaction example MUST contain exactly one HTTP request and one HTTP response message.
+### Enum Element
 
-##### Properties
+Type with the domain of any of the types listed in the `enumerations` attribute. Also called tagged union or Σ-type.
```
Comment (pksunkara) [inline docs/element-definitions.md]: Type with the domain of the union of values typed by Elements in the `enumerations` attribute.

File: docs/element-definitions.md [hunk 32]
```diff
@@ -1,1419 +1,1623 @@
...
 
-- `element`: httpTransaction (string, fixed)
+#### Template
+- `element` - `"enum"`
 - `attributes`
-    - `authSchemes` (Array Element[Base API Element]) - An array of authentication and authorization schemes that apply to the transaction
-- `content` (array) - Request and response message pair (tuple).
-    - (Copy) - HTTP Transaction description's copy text.
-    - (HTTP Request Message)
+  - `enumerations` ([Array](#array-element)) - list of Elements which this Element SHALL type
```
Comment (pksunkara) [inline docs/element-definitions.md]: `list of Elements` is enough

File: docs/element-definitions.md [hunk 33]
```diff
@@ -1,1419 +1,1623 @@
...
+  - `typeAttributes` ([Array](#array-element)[[String](#string-element)])
+    - `fixed` ([String](#string-element)) - Elements in `enumerations` SHALL be interpreted `fixed`.
+  - `validation` - _reserved for future use_
+  - `samples` ([Array](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (Element) - default value for this Element; type of `default` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Bad linking.

File: docs/element-definitions.md [hunk 34]
```diff
@@ -1,1419 +1,1623 @@
...
 
-Ordered array of HTTP header-fields.
-
-##### Properties
-
-- `element`: httpHeaders (string, fixed)
+```
 
-##### Example
+Type Element matching a specific string and numbers (JSON `"Hello world!"` or `number`).
```
Comment (pksunkara) [inline docs/element-definitions.md]: all numbers

File: docs/element-definitions.md [hunk 35]
```diff
@@ -1,1419 +1,1623 @@
...
 
-        If not set, the `href` attribute which was used to resolve the target
-        URI of the parent transition SHOULD be used to resolve the URI of
-        the http request.
+#### Template
+- `element` - `"extend"`
+- `attributes`
+  - `validation` - _reserved for future use_
+  - `samples` ([Array](#array-element)) - array of alternative sample values for this Element; type of items in `samples` MUST match the type of this Element
+  - `default` (Element) - default value for this Element; type of `default` MUST match the type of this Element
```
Comment (pksunkara) [inline docs/element-definitions.md]: Let's remove the `attributes` from here.

File: docs/element-definitions.md (hunk 18 above)
Comment (tjanc) [reply docs/element-definitions.md]: Further applies the `fixed` type attribute on items declaring a non-sample default.

File: docs/element-definitions.md (hunk 22 above)
Comment (tjanc) [reply docs/element-definitions.md]: finite unordered set

File: docs/element-definitions.md [hunk 36]
```diff
@@ -1,1419 +1,1623 @@
...
-##### Properties
+Type with domain of _merged_ Elements specified in `content`. All entries in `content` MUST type the same data structure type. [Ref Elements](#ref-element) encountered in `content` are dereferenced before merging.
 
-- `element`: httpRequest (string, fixed)
-- `attributes`
-    - `method` ([String Element][]) - HTTP request method. The method value SHOULD be inherited from a parent transition if it is unset.
-    - `href` (Templated Href) - URI Template for this HTTP request.
+Merging SHALL be defined based on the type of entries in `content` as follows:
+- [Array Element](#array-element) - array concatenation
+- [Object Element](#object-element) - property concatenation; if duplicit property keys are encountered during concatenation, all but the last SHALL be discarded; tooling SHOULD emit a warning in such a case.
```
[the remark this answers is not in the document]
Comment (tjanc) [reply docs/element-definitions.md]: `set union` on Object Element

**Pull request merged**